analysis/convex.lean

```diff
@@ -0,0 +1,788 @@
+by apply convex_halfspace_lt _ (is_linear_map.mk complex.add_im complex.smul_im)
+
+lemma convex_halfspace_im_le (r : ℝ): convex {c : ℂ | c.im ≤ r} :=
+by apply convex_halfspace_le _ (is_linear_map.mk complex.add_im complex.smul_im)
+
+lemma convex_halfspace_im_gt (r : ℝ): convex {c : ℂ | r < c.im } :=
+by apply convex_halfspace_gt _ (is_linear_map.mk complex.add_im complex.smul_im)
+
+lemma convex_halfspace_im_lge (r : ℝ): convex {c : ℂ | r ≤ c.im} :=
+by apply convex_halfspace_ge _ (is_linear_map.mk complex.add_im complex.smul_im)
+
+lemma convex_sum {γ : Type*} [decidable_eq γ] (hA : convex A) (z : γ → α) (s : finset γ) :
+∀ a : γ → ℝ, s.sum a = 1 → (∀ i ∈ s, 0 ≤ a i) → (∀ i ∈ s, z i ∈ A) → s.sum (λi, a i • z i) ∈ A :=
+begin
+refine finset.induction _ _ s,
+{ intros _ h_sum,
+  simp at h_sum,
+  exact false.elim h_sum },
+{ intros k s hks ih a h_sum ha hz,
+  by_cases h_cases : s.sum a = 0,
+  { have hak : a k = 1,
+      by rwa [finset.sum_insert hks, h_cases, add_zero] at h_sum,
+    have ha': ∀ i ∈ s, 0 ≤ a i,
+      from λ i hi, ha i (finset.mem_insert_of_mem hi),
+    have h_a0: ∀ i ∈ s, a i = 0,
+      from (finset.sum_eq_zero_iff_of_nonneg ha').1 h_cases,
+    have h_az0: ∀ i ∈ s, a i • z i = 0,
+    begin
+      intros i hi,
+      rw h_a0 i hi,
+      exact zero_smul _ (z i)
+    end,
+    show finset.sum (insert k s) (λ (i : γ), a i • z i) ∈ A,
+    begin
+      rw [finset.sum_insert hks, hak, finset.sum_eq_zero h_az0],
+      simp,
+      exact hz k (finset.mem_insert_self k s)
+    end },
+  { have h_sum_nonneg : 0 ≤ s.sum a ,
+    begin
+      apply finset.zero_le_sum',
+      intros i hi,
+      apply ha _ (finset.mem_insert_of_mem hi)
+    end,
+    have h_div_in_A: s.sum (λ (i : γ), ((s.sum a)⁻¹ * a i) • z i) ∈ A,
+    begin
+      apply ih,
+      { rw finset.mul_sum.symm,
+        exact division_ring.inv_mul_cancel h_cases },
+      { intros i hi,
+        exact zero_le_mul (inv_nonneg.2 h_sum_nonneg) (ha i (finset.mem_insert_of_mem hi))},
+      { intros i hi,
+        exact hz i (finset.mem_insert_of_mem hi) }
+    end,
+    have h_sum_in_A: a k • z k
+      + finset.sum s a • finset.sum s (λ (i : γ), ((finset.sum s a)⁻¹ * a i) • z i)
+      ∈ A,
+    begin
+      apply hA,
+      exact hz k (finset.mem_insert_self k s),
+      exact h_div_in_A,
+      exact ha k (finset.mem_insert_self k s),
+      exact h_sum_nonneg,
+      rw (finset.sum_insert hks).symm,
+      exact h_sum
+    end,
+    show finset.sum (insert k s) (λ (i : γ), a i • z i) ∈ A,
+    begin
+      rw finset.sum_insert hks,
+      rw finset.smul_sum at h_sum_in_A,
+      simp [smul_smul, (mul_assoc (s.sum a) _ _).symm] at h_sum_in_A,
+      conv
+      begin
+        congr,
+        congr,
+        skip,
+        congr, skip, funext,
+        rw (one_mul (a _)).symm,
+        rw (field.mul_inv_cancel h_cases).symm,
+      end,
+      exact h_sum_in_A
+    end }
+}
+end
+
+lemma convex_sum_equiv [decidable_eq α]:
```

Comment:
```suggestion
lemma convex_sum_equiv :
```
Add `local attribute [instance] classical.prop_decidable`, but don't add `decidable_eq` at a theorem when it is only used in the proof and not in the statement.